### ISSUE TITLE
[Roles Page] page load is very slow

### ISSUE THREAD

**Issue by aseembansal-gogo** [apache/superset#15854]
On superset 1.1.0 roles list page loading is slow. It takes approx. 16 seconds to load. From roles listing page if I click edit to edit a role then it takes 18 to 20seconds for the edit page to open up. This makes working with roles very difficult.

Viewing user list takes just 2s in comparison.

### Expected results

Roles related pages to load in <5seconds.

### Actual results

Roles related pages load in >15seconds.

#### Screenshots
See developer tools
![image](https://user-images.githubusercontent.com/69132220/126741543-fa37fc48-a80d-416d-9375-27cd4db2c58a.png)

#### How to reproduce the bug

1. Go to Roles list page
2. Edit a role

### Environment

(please complete the following information):

- superset version: `superset version` 1.1.0
- python version: `python --version` Python 3.7.9
- node.js version: `node -v` NA. docker images does not have node

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x]  I did not upgrade to latest release but I searched latest release notes and there isn't anything related to slowness
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by aseembansal-gogo** [comment]
The UI works. It is just slow. The DB is not slow as I am able to query the same tables (`ab_role` etc.) very fast via sql lab. There must be something in the code itself which is causing this slowness when the page is being loaded.

**Comment by grumbler** [comment]
I have the same with the Superset 1.4.2

**Comment by hungpv29** [comment]
I have the same with the Superset 1.5.1. When go to List Roles Page, it calls `Postgres ab_view_menu select ` **44900** times
![image](https://user-images.githubusercontent.com/64894530/196526640-770cc7ba-3370-409a-ae80-fbe99de55728.png)

Query:
```
SELECT ab_view_menu.name AS ab_view_menu_name FROM ab_view_menu JOIN ab_permission_view ON ab_view_menu.id = ab_permission_view.view_menu_id JOIN ab_permission ON ab_permission.id = ab_permission_view.permission_id JOIN ab_permission_view_role ON ab_permission_view.id = ab_permission_view_role.permission_view_id JOIN ab_role ON ab_role.id = ab_permission_view_role.role_id JOIN ab_user_role ON ab_role.id = ab_user_role.role_id JOIN ab_user ON ab_user.id = ab_user_role.user_id WHERE ab_user.id = %(id_1)s AND ab_permission.name = %(name_1)s
```

**Comment by cristian-cevasco** [comment]
Hey, folks! Have anyone found a workaround to this? I'm facing the same with version 2.0.1 and 2.1.0.

**Comment by happykust** [comment]
Hi all! Let me bring some clarity. This problem does not come from Superset, but from the Flask AppBuilder library, on which our favorite product is based. I created just recently there a new issue about this. With the team we are now excavating this problem, because we get up to 5-10 minutes of page load 🤯

https://github.com/dpgaspar/Flask-AppBuilder/issues/2073

**Comment by tirkarthi** [comment]
There was similar issue in Airflow where roles and users page was slow https://github.com/apache/airflow/issues/31340. The fix was to remove certain columns from filter with foreign key reference that causes queries to increase. In Airflow case created_by, changed_by were present in filter widget referencing to users and large number of users increased the the time it takes to load all users. Relevant flask-appbuilder issues

https://github.com/dpgaspar/Flask-AppBuilder/issues/796
https://github.com/dpgaspar/Flask-AppBuilder/issues/875

**Comment by tirkarthi** [comment]
If filtering by permission and users is not needed then please try adding `RoleModelView.search_exclude_columns = ["permissions", "users"]` in below code that should help with page load 

https://github.com/apache/superset/blob/c53b2499988b5aeec96b1064f86703a3891e0efc/superset/security/manager.py#L137-L139

**Comment by happykust** [comment]
@tirkarthi this solution is already in the Superset codebase and it doesn't work in the same way that people's solutions don't with modifying or cutting out the search.
In the `query` function, in the `SQLAInterface` class, where the query itself occurs, there are no external link attachments, which, presumably, at the time the template is rendered, causes a large number of queries.

https://github.com/dpgaspar/Flask-AppBuilder/blob/7f194cfd28099688e0a398ca7b1317aafa804c92/flask_appbuilder/models/sqla/interface.py#L458-L507
`    def query(
        self,
        filters: Optional[Filters] = None,
        order_column: str = "",
        order_direction: str = "",
        page: Optional[int] = None,
        page_size: Optional[int] = None,
        select_columns: Optional[List[str]] = None,
        outer_default_load: bool = False,
    ) -> Tuple[int, List[Model]]:
        """
        Returns the results for a model query, applies filters, sorting and pagination


        :param filters: A Filter class that contains all filters to apply
        :param order_column: name of the column to order
        :param order_direction: the direction to order <'asc'|'desc'>
        :param page: the current page
        :param page_size: the current page size
        :param select_columns: A List of columns to be specifically selected
        on the query. Supports dotted notation.
        :param outer_default_load: If True, the default load for outer joins will be
            applied. This is useful for when you want to control
            the load of the many-to-many relationships at the model level.
            we will apply:
             https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.Load.defaultload
        :return: A tuple with the query count (non paginated) and the results
        """
        if not self.session:
            raise InterfaceQueryWithoutSession()
        query = self.session.query(self.obj)


        count = self.query_count(query, filters, select_columns)
        query = self.apply_all(
            query,
            filters,
            order_column,
            order_direction,
            page,
            page_size,
            select_columns,
        )
        query_results = query.all()


        result = []
        for item in query_results:
            if hasattr(item, self.obj.__name__):
                result.append(getattr(item, self.obj.__name__))
            else:
                return count, query_results
        return count, result`

**Comment by Paulo456** [comment]
search_exclude_columns helps to fast load but permissions - absence
problem solves with adding to find_roles_permission_view_menus and exist_permission_on_roles
this option:
.options(contains_eager(self.permissionview_model.view_menu))
then count query to check menu_view will be lower

bigger problem in representation:
class PermissionView(Model):
....
return str(self.permission).replace("_", " ") + " on " + str(self.view_menu)

from this point we have big amount of queries to two tables ab_view_menu and ab_permission

**Comment by Paulo456** [comment]
in log sqlalchemy looks good if u change model PermissionView to this:
- permission = relationship("Permission")
+ permission = relationship("Permission", lazy="joined")
view_menu_id = Column(Integer, ForeignKey("ab_view_menu.id"))
- view_menu = relationship("ViewMenu")
+ view_menu = relationship("ViewMenu", lazy="joined")

**Comment by Paulo456** [comment]
if u have 50k+ virtual datasets - to edit very slow in frontend superset.
to change this - check only phis datasets for access and parse virt dataset for real tables, like sqllab.


u need to update security manager superset for check real tables and set sql updates (example):

`ALTER TABLE public.ab_view_menu ADD "sql" boolean NULL;

UPDATE public.ab_view_menu SET "sql"=false where 1=1;

UPDATE public.ab_view_menu
SET "sql" = true
from tables t
WHERE t.perm = public.ab_view_menu."name" 
and (t."sql" is not null or t."sql" != '');`

`class PermissionView(Model):
    __tablename__ = "ab_permission_view"
    __table_args__ = (UniqueConstraint("permission_id", "view_menu_id"),)
    id = Column(Integer, Sequence("ab_permission_view_id_seq"), primary_key=True)
    permission_id = Column(Integer, ForeignKey("ab_permission.id"))
    permission = relationship("Permission", lazy="joined")
    view_menu_id = Column(Integer, ForeignKey("ab_view_menu.id"))
    +view_menu = relationship(
        ViewMenu,
        primaryjoin="""and_(
                ViewMenu.id==PermissionView.view_menu_id,
                or_(ViewMenu.sql==False, ViewMenu.sql == None )
                )""",
        innerjoin=True,
        lazy="joined",
    )
....
class ViewMenu(Model):
    __tablename__ = "ab_view_menu"
    id = Column(Integer, Sequence("ab_view_menu_id_seq"), primary_key=True)
    name = Column(String(250), unique=True, nullable=False)
   + sql = Column(Boolean)`

**Comment by Paulo456** [comment]
also add to set_perm
            +if hasattr(target, 'sql') and target.sql:
             +   return

            if not pv and permission and view_menu:
                permission_view_table = (
                    self.permissionview_model.__table__  # pylint: disable=no-member
                )
                connection.execute(
                    permission_view_table.insert().values(
                        permission_id=permission.id, view_menu_id=view_menu.id
                    )
                )

**Comment by ambarishp** [comment]
I am facing this issue in version 3.1.0. The Role edit request taking more that minute. Any fix or work around available

**Comment by Paulo456** [comment]
You can set this to FAB security model
- permission = relationship("Permission")
+ permission = relationship("Permission", lazy="joined")

**Comment by happykust** [comment]
Yup)

File: flask_appbuilder/security/sqla/models.py
Line 76

```
permission = relationship("Permission", lazy="joined")
view_menu = relationship("ViewMenu", lazy="joined")
```

**Comment by rusackas** [comment]
@happykust @ambarishp I'm a little lost in all this now 😅  Is there something that needs to be done to close the loop here (adding documentation, a PR to fix things, etc.?)

**Comment by Paulo456** [comment]
@rusackas  I think the problem is that the superset is not designed for the number of dataset entities greater than 50 thousand. More precisely, the problem is not in the superset itself, but in the FAB model. The current solution is just a workaround.

**Comment by happykust** [comment]
@rusackas The problem is solely in the Flask-AppBuilder library, in the file described above. Its author apparently does not want to fix this; there have already been a lot of discussions about this. In our company, we forked FAB, fixed it and are enjoying life.

**Comment by rusackas** [comment]
CC @dpgaspar

**Comment by dpgaspar** [comment]
@happykust I'm sorry to ear that, can you point to any current/past issue/PR on flask-appbuilder or open a new one please, so we can visit this.

**Comment by happykust** [comment]
@dpgaspar 

https://github.com/dpgaspar/Flask-AppBuilder/issues/2073

**Comment by happykust** [comment]
@dpgaspar I've been adding an issue for quite some time now. I switched it to closed status because of the solution found for me. Will you have a chance to fix this? Or are there any pitfalls when using the solution written above?

**Comment by happykust** [comment]
@dpgaspar Thank you very much for the fixes!

**Comment by pashtet04** [comment]
Which version of Superset has this fix?

The issue is still in 3.1.1.
And I couldn't find where to fix it manually in the container image because there is nothing in src except the `migrations` directory

```
superset@superset-1661616798-779bfb6d69-vxsqw:/app$ grep -ri 'permission = relationship' .
./superset/migrations/versions/2020-09-24_12-04_3fbbc6e8d654_fix_data_access_permissions_for_virtual_.py:    permission = relationship("Permission")
./superset/migrations/versions/2022-05-18_16-07_e786798587de_delete_none_permissions.py:    permission = relationship("Permission")
./superset/migrations/shared/security_converge.py:    permission = relationship("Permission")
```

**Comment by ohappykust** [comment]
@pashtet04 

Yep, Superset 3.1.1 doesn't have this fixes because of use FAB version 4.3.11, not 4.4.1.
You should update your Superset instance to 3.1.2 version.
